var/spack/repos/builtin/packages/gcc/package.py

```diff
@@ -82,60 +95,111 @@ class Gcc(AutotoolsPackage):
     # depends_on('autogen@5.5.4:', type='test')
     # depends_on('guile@1.4.1:', type='test')
 
+    # See https://golang.org/doc/install/gccgo#Releases
+    provides('golang', when='languages=go')
+    # 'when' does not currently support multiple parts of a spec.
+    # See https://github.com/LLNL/spack/pull/4151
+    # provides('golang',        when='languages=go @4.6:')
+    # provides('golang@:1',     when='languages=go @4.7.1:')
+    # provides('golang@:1.1',   when='languages=go @4.8:')
+    # provides('golang@:1.1.2', when='languages=go @4.8.2:')
+    # provides('golang@:1.2',   when='languages=go @4.9:')
+    # provides('golang@:1.4',   when='languages=go @5:')
+    # provides('golang@:1.6.1', when='languages=go @6:')
+    # provides('golang@:1.8',   when='languages=go @7:')
+
+    # For a list of valid languages for a specific release,
+    # run the following command in the GCC source directory:
+    #    $ grep ^language= gcc/*/config-lang.in
+    # See https://gcc.gnu.org/install/configure.html
+
+    # Support for processing BRIG 1.0 files was added in GCC 7
+    # BRIG is a binary format for HSAIL:
+    # (Heterogeneous System Architecture Intermediate Language).
+    # See https://gcc.gnu.org/gcc-7/changes.html
+    conflicts('languages=brig', when='@:6')
+
+    # BRIG does not seem to be supported on macOS
+    conflicts('languages=brig', when='platform=darwin')
+
+    # GCC 4.8 added a 'c' language. I'm sure C was always built,
+    # but this is the first version that accepts 'c' as a valid language.
+    conflicts('languages=c', when='@:4.7')
+
+    # GCC 4.6 added support for the Go programming language.
+    # See https://gcc.gnu.org/gcc-4.6/changes.html
+    conflicts('languages=go', when='@:4.5')
+
+    # Go is not supported on macOS
+    conflicts('languages=go', when='platform=darwin')
+
+    # The GCC Java frontend and associated libjava runtime library
+    # have been removed from GCC as of GCC 7.
+    # See https://gcc.gnu.org/gcc-7/changes.html
+    conflicts('languages=java', when='@7:')
+
+    # GCC 5 added the ability to build GCC as a Just-In-Time compiler.
+    # See https://gcc.gnu.org/gcc-5/changes.html
+    conflicts('languages=jit', when='@:4')
+
     if sys.platform == 'darwin':
+        patch('darwin/gcc-7.1.0-headerpad.patch', when='@5:')
+        patch('darwin/gcc-6.1.0-jit.patch', when='@5:')
```

Comment:
Confirmed that these patches apply for all versions of GCC 5-7, and are in fact required in order to build the Just-In-Time compiler on macOS. Not sure if they are darwin specific or clang specific though.